src/storage/txn/scheduler.rs

```diff
@@ -919,40 +973,78 @@ impl Scheduler {
+                .entry(HashableContext(ctx))
+                .or_insert_with(Vec::new);
+            group.push(cid);
         }
     }
 
     pub fn run(&mut self, receiver: Receiver<Msg>) -> Result<()> {
+        let mut msgs = Vec::with_capacity(CMD_BATCH_SIZE);
         loop {
             let msg = box_try!(receiver.recv());
-            match msg {
-                Msg::Quit => return Ok(()),
-                Msg::RawCmd { cmd, cb } => self.on_receive_new_cmd(cmd, cb),
-                Msg::SnapshotFinished { cid, cb_ctx, snapshot } => {
-                    self.on_snapshot_finished(cid, cb_ctx, snapshot)
+            msgs.push(msg);
+            loop {
+                match receiver.try_recv() {
+                    Ok(msg) => {
+                        msgs.push(msg);
+                        if msgs.len() >= CMD_BATCH_SIZE {
+                            break;
+                        }
+                    }
+                    Err(TryRecvError::Empty) => break,
+                    _ => (),
                 }
-                Msg::ReadFinished { cid, pr } => self.on_read_finished(cid, pr),
-                Msg::WritePrepareFinished { cid, cmd, pr, to_be_write } => {
-                    self.on_write_prepare_finished(cid, cmd, pr, to_be_write)
+            }
+
+            for msg in msgs.drain(..) {
+                match msg {
+                    Msg::Quit => return Ok(()),
+                    Msg::RawCmd { cmd, cb } => self.on_receive_new_cmd(cmd, cb),
+                    Msg::SnapshotFinished { cids, cb_ctx, snapshot } => {
+                        self.on_snapshot_finished(cids, cb_ctx, snapshot)
+                    }
+                    Msg::ReadFinished { cid, pr } => self.on_read_finished(cid, pr),
+                    Msg::WritePrepareFinished { cid, cmd, pr, to_be_write } => {
+                        self.on_write_prepare_finished(cid, cmd, pr, to_be_write)
+                    }
+                    Msg::WritePrepareFailed { cid, err } => self.on_write_prepare_failed(cid, err),
+                    Msg::WriteFinished { cid, pr, result, .. } => {
+                        self.on_write_finished(cid, pr, result)
+                    }
                 }
-                Msg::WritePrepareFailed { cid, err } => self.on_write_prepare_failed(cid, err),
-                Msg::WriteFinished { cid, pr, result, .. } => {
-                    self.on_write_finished(cid, pr, result)
+            }
+
+            let mut total = 0;
+            if !self.grouped_cmds.as_ref().unwrap().is_empty() {
+                let m = self.grouped_cmds.take().unwrap();
+                for (ctx, cids) in m {
+                    if !cids.is_empty() {
+                        total += cids.len();
+                        self.get_snapshot(&ctx.0, cids);
+                    }
                 }
+                self.grouped_cmds = Some(HashMap::with_capacity(CMD_BATCH_SIZE));
             }
+            BATCH_COMMANDS_GAUGE.set(total as f64);
```

Comment:
I think the meaningful metrics is whether we always get real batches. So every `cids` should be observed using histogram.